src/traits/wf.md

```diff
@@ -1,11 +1,444 @@
+associated type definition in the trait declaration, *except* that type
+parameters from the trait are substituted with values provided by the impl
+(see example below). You cannot add new where clauses. You may omit to write
+the where clauses if you want to emphasize the fact that you are actually not
+relying on them.
+
+Some examples to illustrate that:
+```rust,ignore
+trait Foo<T> {
+    type Assoc where T: Clone;
+}
+
+struct OnlyClone<T: Clone> { ... }
+
+impl<U> Foo<Option<U>> for () {
+    // We substitute type parameters from the trait by the ones provided
+    // by the impl, that is instead of having a `T: Clone` where clause,
+    // we have an `Option<U>: Clone` one.
+    type Assoc = OnlyClone<Option<U>> where Option<U>: Clone;
+}
+
+impl<T> Foo<T> for i32 {
+    // I'm not using the `T: Clone` where clause from the trait, so I can
+    // omit it.
+    type Assoc = u32;
+}
+
+impl<T> Foo<T> for f32 {
+    type Assoc = OnlyClone<Option<T>> where Option<T>: Clone;
+    //                                ^^^^^^^^^^^^^^^^^^^^^^
+    //                                this where clause does not exist
+    //                                on the original trait decl: illegal
+}
+```
+
+So where clauses on associated types work *exactly* like where clauses on
+trait methods: in an impl, we must substitute the parameters from the traits
+with values provided by the impl, we may omit them if we don't need them, but
+we cannot add new where clauses.
+
+Now let's see the generated goal for this general impl:
+```text
+forall<P1...> {
+    if (FromEnv(WC_impl), FromEnv(InputTypes(SomeType<A2...>))) {
+        WellFormed(SomeType<A2...>: Trait<A1...>) &&
+            WellFormed(InputTypes(WC_impl)) &&
+
+            forall<P2...> {
+                if (FromEnv(WC_assoc)) {
+                    WellFormed(SomeValue<A3...>: Bounds_assoc) &&
+                        WellFormed(InputTypes(SomeValue<A3...>))
+                }
+            }
+    }
+}
+```
+
+Here is the most complex goal. As always, first, assuming that
+the various where clauses hold, we prove that every type appearing in the impl
+is well-formed, ***except*** types appearing in the receiver type
+`SomeType<A2...>`. Instead, we *assume* that those types are well-formed
+(hence the `if (FromEnv(InputTypes(SomeType<A2...>)))` condition). This is
+part of the implied bounds proposal, so that we can rely on the bounds
+written on the definition of the `SomeType<A2...>` type (and that we don't
+need to repeat those bounds).
+
+Next, assuming that the where clauses on the impl `WC_impl` hold and that the
+input types of `SomeType<A2...>` are well-formed, we prove that
+`WellFormed(SomeType<A2...>: Trait<A1...>)` hold. That is, we want to prove
+that `SomeType<A2...>` verify all the where clauses that might transitively
+come from the `Trait` definition (see
+[this subsection](./implied-bounds.md#co-inductiveness-of-wellformed)).
```

Comment:
Actually I was trying to factor out all the `WellFormed(InputTypes(...))` in the previous paragraph (« assuming that the various where clauses hold, we prove that every type appearing in the impl is WF»)

Reply:
Oh, I missed that, sorry. Maybe update the goal clauses to match the order that they are mentioned in the text?

Also, it does say something to the effect of "assuming that the various where clauses hold" twice, this might be slightly confusing.

Reply:
Yeah I think when writing that paragraph I wasn’t entierely convinced by my formulation either :)

I’ll separate the if goals in two parts yes, it should be clearer.